### DISCUSSION TITLE
Any way to get the entire time series matrix including signals, trades and position?

### DISCUSSION THREAD

**Discussion by felonfusk** [kernc/backtesting.py#1024] [Q&A] [answered]
Often times, I find it necessary to combine the market data with indicators / signals, trades and positions into one huge DataFrame.
This allows me to export it into .CSV and compare with the same columns done with Excel formula for debugging.
For example: the first couple of columns are OHLCV data, followed by some indicators (MACD, RSI, etc.), then trades and positions, and finally the portfolio equity.
I couldn't find where the trades, positions and equity time series are stored, and I wonder whether the signals time series are stored at all.
Thanks

**Comment by peacerider** [comment, marked as answer]
Is this data useful for you?

*output._equity_curve
*output._trades

```
from backtesting import Backtest, Strategy
from backtesting.lib import crossover

from backtesting.test import SMA, GOOG


class SmaCross(Strategy):
    n1 = 10
    n2 = 20

    def init(self):
        close = self.data.Close
        self.sma1 = self.I(SMA, close, self.n1)
        self.sma2 = self.I(SMA, close, self.n2)

    def next(self):
        if crossover(self.sma1, self.sma2):
            self.buy()
        elif crossover(self.sma2, self.sma1):
            self.sell()


bt = Backtest(GOOG, SmaCross,
              cash=10000, commission=.002,
              exclusive_orders=True)

output = bt.run()
bt.plot()

print(output._equity_curve)
print(output._trades)
```

**Reply by felonfusk** [reply]
Perfect. Thank you

**Reply by felonfusk** [reply]
Sorry do you also happen to know where to get the indicator time series?
Like the sample strategy in the front page, there are 2 SMAs: SMA10 and SMA20.
They are in the chart but I don't see how to access them.

**Reply by peacerider** [reply]
I think that you need to calculate the indicator time series by yourself.
(I am beginer of backtesting.py. There may be other, better ways.)

Built in SMA function I used in the front page is as below.
https://github.com/kernc/backtesting.py/blob/master/backtesting/test/__init__.py
```
def SMA(arr: pd.Series, n: int) -> pd.Series:
    """
    Returns `n`-period simple moving average of array `arr`.
    """
    return pd.Series(arr).rolling(n).mean()
```
So,  you can get the SMA indicator time series as below.

sma10 = pd.Series(GOOG["Close"]).rolling(10).mean()
sma20 = pd.Series(GOOG["Close"]).rolling(20).mean()

```
import pandas as pd
from backtesting.test import SMA, GOOG

sma10 = pd.Series(GOOG["Close"]).rolling(10).mean()
sma20 = pd.Series(GOOG["Close"]).rolling(20).mean()

print(sma10)
print(sma20)
```
